src/app/encoder.cpp

```diff
@@ -0,0 +1,170 @@
+/**
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "chip-zcl-zpro-codec.h"
+#include <assert.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <support/BufBound.h>
+#include <support/logging/CHIPLogging.h>
+
+using namespace chip;
+extern "C" {
+
+static uint16_t doEncodeApsFrame(BufBound & buf, uint16_t profileID, uint16_t clusterId, uint8_t sourceEndpoint,
+                                 uint8_t destinationEndpoint, EmberApsOption options, uint16_t groupId, uint8_t sequence,
+                                 uint8_t radius)
+{
+
+    uint8_t control_byte = 0;
+    buf.Put(control_byte); // Put in a control byte
+    buf.PutLE16(profileID);
+    buf.PutLE16(clusterId);
+    buf.Put(sourceEndpoint);
+    buf.Put(destinationEndpoint);
+    buf.PutLE(options, sizeof(EmberApsOption));
+    buf.PutLE16(groupId);
+    buf.Put(sequence);
+    buf.Put(radius);
+
+    if (!buf.Fit() && buf.Size() > 0)
```

Comment:
I'm not a big fan of treating "size == 0 means measure", especially since that's not what the API docs say.  If the caller passes a non-null buffer and a 0 buffer size, they will get an incorrect (per current API docs) return from this function, no?

We could change the API contract to use size instead of buffer as the "we are measuring" case, I guess, but it feels like it's too easy to accidentally land in the measuring case from someplace that just runs out of buffer in a natural way and then suddenly gets an unexpected answer.

Maybe `doEncodeApsFrame` should just get an explicit boolean for whether we are measuring, and most callers will pass false, while `encodeApsFrame` passes `!buffer` or something?

Reply:
Fair will fix.